Kubernetes compatibility with Statefulset

romdum opened the pull request:
Hi !

This pull request add the possibility of scale down and up Statefulset in addtition to Deployment when Kubernetes mode is enabled.

Some precisions : 
* I update the package `k8s.io/client-go` from `v0.20.6` to `v0.22.3` to be able to use Statefulset.Status.AvailableReplicas
* In the condition below I replaced AvailableReplicas by ReadyReplicas because AvailableReplicas is always at 0 with Statefulset
```go
if d.Status.ReadyReplicas > 0 {
	return true
}
```
* I tested this PR on k3s cluster with kubernetes 1.21, traefik 2.5.3, I created two middlewares on the same ingress like below : 
```yml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: nextcloud
  namespace: nextcloud
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt
    traefik.ingress.kubernetes.io/router.entrypoints: websecure
    traefik.ingress.kubernetes.io/router.middlewares: nextcloud-container-on-demand-mariadb@kubernetescrd,nextcloud-container-on-demand@kubernetescrd
    traefik.ingress.kubernetes.io/router.tls: "true"
spec:
  rules:
    - host: nextcloud.dns
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name:  nextcloud
                port:
                  number: 8080
  tls:
  - secretName: certificate
    hosts:
      - nextcloud.dns
---
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: container-on-demand
  namespace: nextcloud
spec:
  plugin:
    traefik-ondemand-plugin:
      name: deployment_nextcloud_nextcloud_1
      serviceUrl: 'http://traefik-ondemand-service.kube-system.svc:10000'
      timeout: 30m
---
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: container-on-demand-mariadb
  namespace: nextcloud
spec:
  plugin:
    traefik-ondemand-plugin:
      name: statefulset_nextcloud_nextcloud-mariadb_1
      serviceUrl: 'http://traefik-ondemand-service.kube-system.svc:10000'
      timeout: 30m
```

acouvreur commented:
Maybe you can take a look @mschneider82 ?

acouvreur commented:
Can you provide an example with your specification in the same format as : 

https://github.com/acouvreur/traefik-ondemand-plugin/tree/main/examples/kubernetes

As simple "how to" guide from the very beginning with all resources needed to test it.

romdum commented:
Ok so I have the following Kubernetes resources : 
A deployment for Nextcloud app (extracted and simplified from helm chart) : 
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/component: app
    app.kubernetes.io/instance: nextcloud
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: nextcloud
    helm.sh/chart: nextcloud-2.8.3
  name: nextcloud
  namespace: nextcloud
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/component: app
      app.kubernetes.io/instance: nextcloud
      app.kubernetes.io/name: nextcloud
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: app
        app.kubernetes.io/instance: nextcloud
        app.kubernetes.io/name: nextcloud
    spec:
      containers:
      - env:
        - name: MYSQL_HOST
          value: nextcloud-mariadb
          #  some  others env var ...
        image: nextcloud:21.0.5-apache
        imagePullPolicy: IfNotPresent
        name: nextcloud
        ports:
        - containerPort: 80
          name: http
          protocol: TCP
        volumeMounts:
        - mountPath: /var/www/
          name: nextcloud-data
          subPath: root
         # some others mount
      restartPolicy: Always
      volumes:
      - name: nextcloud-data
        persistentVolumeClaim:
          claimName: nextcloud-nextcloud
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app.kubernetes.io/component: app
    app.kubernetes.io/instance: nextcloud
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: nextcloud
    helm.sh/chart: nextcloud-2.8.3
  name: nextcloud
  namespace: nextcloud
spec:
  ports:
  - name: http
    port: 8080
    targetPort: http
  selector:
    app.kubernetes.io/component: app
    app.kubernetes.io/name: nextcloud
```
After I have a Statefulset which deploy the Maria DB for my nextcloud app : 
```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: nextcloud-mariadb
  namespace: nextcloud
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nextcloud-mariadb
  serviceName: nextcloud-mariadb
  template:
    metadata:
      labels:
        app: nextcloud-mariadb
    spec:
      containers:
        - name: nextcloud-mariadb
          image: mariadb:10.3
          imagePullPolicy: Always
          ports:
            - name: mysql
              containerPort: 3306
          envFrom:
            - secretRef:
                name: database-secret
          volumeMounts:
            - mountPath: "/var/lib/mysql"
              name: data-nextcloud-mariadb
              subPath: mysql
  volumeClaimTemplates:
    - metadata:
        name: data-nextcloud-mariadb
      spec:
        accessModes: [ "ReadWriteOnce" ]
        storageClassName: "longhorn"
        resources:
          requests:
            storage: 5Gi
---
apiVersion: v1
kind: Service
metadata:
  name: nextcloud-mariadb
spec:
  selector:
    app: nextcloud-mariadb
  ports:
    - name: mysql
      port: 3306
      targetPort: 3306
  type: ClusterIP
```
Finally I have an Ingress to expose my app, in this ingress I have two middleware, one for the nextcloud deployment and one for the mariadb server : 
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: nextcloud
  namespace: nextcloud
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt
    traefik.ingress.kubernetes.io/router.entrypoints: websecure
    traefik.ingress.kubernetes.io/router.middlewares: nextcloud-container-on-demand-mariadb@kubernetescrd,nextcloud-container-on-demand@kubernetescrd
    traefik.ingress.kubernetes.io/router.tls: "true"
spec:
  rules:
    - host: nextcloud.dns
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name:  nextcloud
                port:
                  number: 8080
  tls:
  - secretName: certificate
    hosts:
      - nextcloud.dns
---
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: container-on-demand
  namespace: nextcloud
spec:
  plugin:
    traefik-ondemand-plugin:
      name: deployment_nextcloud_nextcloud_1
      serviceUrl: 'http://traefik-ondemand-service.kube-system.svc:10000'
      timeout: 30m
---
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: container-on-demand-mariadb
  namespace: nextcloud
spec:
  plugin:
    traefik-ondemand-plugin:
      name: statefulset_nextcloud_nextcloud-mariadb_1
      serviceUrl: 'http://traefik-ondemand-service.kube-system.svc:10000'
      timeout: 30m
```

Traefik is deployed exactly like in https://github.com/acouvreur/traefik-ondemand-plugin/tree/main/examples/kubernetes with this parameters in values.yml : 
```yml
# traefik helm values
additionalArguments:
  - "--pilot.token=XXXXX_YOURTOKEN_XXXXXXXXXXXXXXXX"
  - "--experimental.plugins.traefik-ondemand-plugin.modulename=github.com/acouvreur/traefik-ondemand-plugin"
  - "--experimental.plugins.traefik-ondemand-plugin.version=v1.0.1"
  - "--providers.kubernetesingress.allowEmptyServices=true"

experimental:
  plugins:
    enabled: true
```

And traefik-ondemand-service is deploy like in manifest.yml (I just added RBAC for statefulset in PR) : https://github.com/acouvreur/traefik-ondemand-plugin/blob/main/examples/kubernetes/manifests.yml

mschneider82 commented:
This documentation should be included into the commit in the examples directory. 
What happens to PersistantVolumes when it are scaled down to 0 and than back to 1? 
I added a note to a codepart where the old switch statement was nicer than if then else...

romdum commented:
> What happens to PersistantVolumes when it are scaled down to 0 and than back to 1?
When it scale down, the volumeAttachment is removed, and when it scale up the volumeAttachment is recreated, I think that's all.

romdum commented on pkg/scaler/docker_kubernetes.go:
```diff
@@ -104,33 +111,35 @@ func (scaler *KubernetesScaler) ScaleDown(name string) error {
-			log.Infof("Replicas for %s %s in namespace %s are already: 0", config.Kind, config.Name, config.Namespace)
-			return nil
-		}
+	if config.Kind == "deployment" {
```
Yes it's done

acouvreur commented:
:tada: This PR is included in version 1.7.0 :tada:

The release is available on:
- `v1.7.0`
- [GitHub release](https://github.com/acouvreur/traefik-ondemand-service/releases/tag/v1.7.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
